intellij-lsp/src/com/github/gtache/lsp/editor/EditorEventManager.scala

```diff
@@ -1104,42 +1103,39 @@ class EditorEventManager(val editor: Editor, val mouseListener: EditorMouseListe
-                  LOG.warn(e)
-                  wrapper.crashed(e.asInstanceOf[Exception])
-                  new Range(serverPos, serverPos)
-              }
-            } else new Range(serverPos, serverPos)
-
-          } else range
-          val startOffset = DocumentUtils.LSPPosToOffset(editor, corRange.getStart)
-          val endOffset = DocumentUtils.LSPPosToOffset(editor, corRange.getEnd)
-          val isDefinition = DocumentUtils.LSPPosToOffset(editor, loc.getRange.getStart) == startOffset
-          if (ctrlRange != null) ctrlRange.dispose()
-          ctrlRange = CtrlRangeMarker(loc, editor,
-            if (!isDefinition) editor.getMarkupModel.addRangeHighlighter(startOffset, endOffset, HighlighterLayer.HYPERLINK, editor.getColorsScheme.getAttributes(EditorColors.REFERENCE_HYPERLINK_COLOR), HighlighterTargetArea.EXACT_RANGE)
-            else null)
-        }
-      })
+      if (!editor.isDisposed) {
+        val corRange = if (range == null) {
+          val params = new TextDocumentPositionParams(identifier, serverPos)
+          val future = requestManager.documentHighlight(params)
+          if (future != null) {
+            try {
+              val highlights = future.get(DOC_HIGHLIGHT_TIMEOUT, TimeUnit.MILLISECONDS)
+              if (highlights != null) {
+                wrapper.notifySuccess(Timeouts.DOC_HIGHLIGHT)
+                val offset = DocumentUtils.LSPPosToOffset(editor, serverPos)
+                highlights.asScala.find(dh => DocumentUtils.LSPPosToOffset(editor, dh.getRange.getStart) <= offset
+                  && offset <= DocumentUtils.LSPPosToOffset(editor, dh.getRange.getEnd)).fold(new Range(serverPos, serverPos))(dh => dh.getRange)
+              } else new Range(serverPos, serverPos)
+            } catch {
+              case e: TimeoutException =>
+                LOG.warn(e)
+                wrapper.notifyFailure(Timeouts.DOC_HIGHLIGHT)
+                new Range(serverPos, serverPos)
+              case e@(_: java.io.IOException | _: JsonRpcException | _: ExecutionException) =>
+                LOG.warn(e)
+                wrapper.crashed(e.asInstanceOf[Exception])
+                new Range(serverPos, serverPos)
+            }
+          } else new Range(serverPos, serverPos)
+        } else range
+        val startOffset = DocumentUtils.LSPPosToOffset(editor, corRange.getStart)
+        val endOffset = DocumentUtils.LSPPosToOffset(editor, corRange.getEnd)
+        val isDefinition = DocumentUtils.LSPPosToOffset(editor, loc.getRange.getStart) == startOffset
+        if (ctrlRange != null) ctrlRange.dispose()
+        ctrlRange = CtrlRangeMarker(loc, editor,
+          if (!isDefinition) editor.getMarkupModel.addRangeHighlighter(startOffset, endOffset, HighlighterLayer.HYPERLINK, editor.getColorsScheme.getAttributes(EditorColors.REFERENCE_HYPERLINK_COLOR), HighlighterTargetArea.EXACT_RANGE)
+          else null)
```

Comment:
You need to wrap the call to dispose and the addRangeHighlighter into an invokeLater, otherwise we get an exception (only the main thread can modify the GUI), so basically     
`invokeLater(() => {
    if (ctrlRange!=null)...
    if (!isDefinition) ... else null)})`

Reply:
Oh sorry missed that, Thanks for pointing that out! :)

Reply:
I think you put the invokeLater in the wrong method (see other comment) ;)

Reply:
Sure will remove it (line 807). So we are all good, right? :) or am I still missing something?

Reply:
Hahaha nope, it is still missing here (lines 1138-1142)
It should be 
`invokeLater(() => {
    if (ctrlRange != null) ctrlRange.dispose()
    ctrlRange = CtrlRangeMarker(loc, editor, if (!isDefinition) editor.getMarkupModel... else null)
})`

Reply:
Fixed :)